packages/SwingSet/src/vats/timer/vat-timer.js

```diff
@@ -1,98 +1,854 @@
+      if (scheduled) {
+        removeEvent(schedule, scheduled, self);
+        state.scheduled = undefined;
+        reschedule();
+        if (wakeupPromises.has(self)) {
+          wakeupPromises.get(self).reject({ name: 'TimerCancelled' });
+        }
+      }
+    },
+  };
+
+  const makePromiseEvent = defineDurableKind(
+    promiseEventHandle,
+    initPromiseEvent,
+    promiseEventBehavior,
+  );
+
+  // -- Event (repeaters)
+
+  function initRepeaterEvent(startTime, interval, handler, cancelToken) {
+    const scheduled = undefined;
+    const cancelled = false;
+    const state = {
+      startTime,
+      interval,
+      handler,
+      scheduled,
+      cancelled,
+      cancelToken,
+    };
+    return state;
+  }
+
+  const repeaterEventBehavior = {
+    scheduleYourself({ self, state }) {
+      // first time
+      const { startTime, interval, cancelToken } = state;
+      const now = getNow();
+      const when = nextScheduleTime(startTime, interval, now);
+      state.scheduled = when; // cleared if fired/cancelled
+      addEvent(schedule, when, self);
+      if (cancelToken) {
+        addCancel(cancels, cancelToken, self);
+      }
+      reschedule();
+    },
+
+    rescheduleYourself({ self, state }) {
+      const { cancelled, startTime, interval } = state;
+      if (cancelled) {
+        // cancelled while waiting for handler to finish
+        return;
+      }
+      const now = getNow();
+      const when = nextScheduleTime(startTime, interval, now);
+      state.scheduled = when; // cleared if fired/cancelled
+      addEvent(schedule, when, self);
+      reschedule();
+    },
+
+    fired({ self, state }) {
+      const { cancelled, scheduled, handler } = state;
+      state.scheduled = undefined;
+      if (cancelled) {
+        return;
+      }
+      // repeaters stay in "waiting" until their promise resolves,
+      // at which point we either reschedule or cancel
+      E(handler)
+        .wake(toTimestamp(scheduled))
+        .then(
+          _res => self.rescheduleYourself(),
+          _err => self.cancel(),
+        )
+        .catch(err => console.log(`timer repeater error`, err));
+    },
+
+    cancel({ self, state }) {
+      const { scheduled, cancelToken } = state;
+      removeCancel(cancels, cancelToken, self);
+      state.cancelled = true;
+      if (scheduled) {
+        removeEvent(schedule, scheduled, self);
+        state.scheduled = undefined;
+        reschedule();
+      }
+    },
+  };
+
+  const makeRepeaterEvent = defineDurableKind(
+    repeaterEventHandle,
+    initRepeaterEvent,
+    repeaterEventBehavior,
+  );
+
+  // -- more helper functions
+
+  /**
+   * @param {TimestampValue} when
+   * @param {CancelToken} cancelToken
+   * @returns {Promise<Timestamp>}
+   */
+  function wakeAtInternal(when, cancelToken) {
+    const event = makePromiseEvent(when, cancelToken);
+    const { resolve, reject, promise } = makePromiseKit();
+    const controls = harden({ resolve, reject });
```

Comment:
See comment on line 241

```suggestion
    const controls = Far('promise control', { resolve, reject });
```

Reply:
I see it's done. Resolving.